fms_extras/utils/generation.py

```diff
@@ -3,12 +3,344 @@
+        [10, 11, 12, 13],
+    ]
+
+    context_lengths -
+
+    [19, 20, 21, 22, 63, 64, 65, 66]
+
+    Args:
+        cache_data: PagedAttentionCacheData
+            the cache data created after allocation
+
+    Returns:
+    PagedAttentionCacheData
+        The cache data after applying batch expansion
+    """
+    # Set up kv cache metadata for paged memory access over tokens/candidates with shared prefixes
+    context_lengths = cache_data.context_lengths  # bk
+    inflate_factor = (
+        cache_data.query_length
+        if cache_data.unflatten_indices is None
+        else cache_data.unflatten_indices.size(-1)
+    )
+    # no reason to check type here as generation allocation always returns context_lengths
+    context_lengths = context_lengths.unsqueeze(1).expand(  # type: ignore
+        -1, inflate_factor
+    )  # bk n
+    # subtract arange(inp_len) to get lengths for each token in candidates
+    cache_data.context_lengths = (
+        context_lengths.sub(context_lengths.sign().cumsum(1).flip([1]).sub(1))
+        .int()
+        .view(-1)
+    )  # bkn
+    cache_data.block_mapping = cache_data.block_mapping.repeat_interleave(
+        inflate_factor, dim=0
+    )  # bkn n_blocks
+    return cache_data
+
+
+def __prepare_cache_metadata_for_flattening(
+    cache_data: PagedAttentionCacheData,
+    unflat_indices: torch.Tensor,
+    flat_indices: torch.Tensor,
+) -> PagedAttentionCacheData:
+    """Add the proper metadata to the cache required to perform batch flattening/unflattening
+
+    Args:
+        cache_data: PagedAttentionCacheData
+            the paged attention cache data / metadata
+        unflat_indices: torch.Tensor
+            the indices used in performing unflattening prior to storing in the cache
+        flat_indices: torch.Tensor
+            the indices used in performing flattening
+
+    Returns:
+    PagedAttentionCacheData
+        the cache data with proper unflattening metadata included
+    """
+    cache_data.unflatten_indices = unflat_indices
+    cache_data.flatten_indices = flat_indices
+    cache_data.position_ids = apply_index_map(
+        cache_data.position_ids.view(-1), flat_indices
+    )[
+        None,
+    ]
+    cache_data.context_lengths = apply_index_map(
+        cache_data.context_lengths, cache_data.flatten_indices  # type: ignore
+    )  # n'
+    cache_data.block_mapping = apply_index_map(
+        cache_data.block_mapping, cache_data.flatten_indices
+    )  # n' n_blocks
+
+    return cache_data
+
+
+def __perform_correctness_checks(
+    input_ids_unflat: torch.Tensor,
+    next_vals: torch.Tensor,
+    embeds: torch.Tensor,
+    batch_size: int,
+    num_candidates_per_sequence: int,
+    decode_seq_length: int,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Check for the correctness of speculator predictions and get the best candidates as well as the number of tokens
+    correct for each of those candidates
```

Comment:
More precisely, we're getting the indices of the best candidates, plus the number of tokens correct, and the base model output values for that candidate (tokens and embeddings)

Reply:
fixed in latest commit